packages/@react-aria/collections/src/CollectionBuilder.tsx

```diff
@@ -127,22 +127,40 @@ function useCollectionDocument<T extends object, C extends BaseCollection<T>>(cr
 
 const SSRContext = createContext<BaseNode<any> | null>(null);
 
-function useSSRCollectionNode<T extends Element>(Type: string, props: object, ref: ForwardedRef<T>, rendered?: any, children?: ReactNode, render?: (node: Node<T>) => ReactElement) {
+export type CollectionNodeClass<T> = {
+  new (key: Key): CollectionNode<T>,
+  readonly type: string
+};
+
+function createCollectionNodeClass(type: string): CollectionNodeClass<any> {
+  let NodeClass = function (key: Key) {
+    return new CollectionNode(type, key);
+  } as any;
+  NodeClass.type = type;
+  return NodeClass;
```

Comment:
```suggestion
  let NodeClass = class extends CollectionNode<any> {
    static readonly type = type;
    constructor(key: Key) {
      super(type, key);
    }
  };
  return NodeClass;
```
I think this is a little clearer as to what is going on

Reply:
you could memo the creation of the NodeClass as well if you wanted, then all the instances would be match